src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1686,21 +1686,19 @@ private boolean areDisjoint(ClassSymbol ts, ClassSymbol ss) {
                 if (!pairsSeen.add(newPair))
                     return false;
 
-                if (ts.isInterface() && !ss.isInterface() || // case I: one is a class and the other one is an interface
-                    !ts.isInterface() && ss.isInterface()) {
+                if (ts.isInterface() != ss.isInterface()) { // case I: one is a class and the other one is an interface
                     ClassSymbol isym = ts.isInterface() ? ts : ss; // isym is the interface and csym the class
                     ClassSymbol csym = isym == ts ? ss : ts;
                     if (!isSubtype(erasure(csym.type), erasure(isym.type))) {
                         if (csym.isFinal()) {
                             return true;
                         } else if (csym.isSealed()) {
                             return areDisjoint(isym, csym.getPermittedSubclasses());
-                        } else if (isClassFreelyExtensible(csym) && isym.isSealed()) {
+                        } else if (isym.isSealed()) {
```

Comment:
where did the freely extensible check go? That part is in the spec, but there's no code for it?

Reply:
Jan suggested in a previous comment that: `isClassFreelyExtensible == !csym.isSealed() && !csym.isFinal()`, which I think is correct. At this point we have checked that the class is not final and not sealed so if the interface is sealed we do the check for disjoint-ness between the class and the interface's permitted subclasses

Reply:
Yes, I suggested that. The idea being that even if a class has a sealed supertype, then it must be one of `final`, `sealed` and `non-sealed` itself (or a compile-time error is produced). So, basically all classes that are not freely extensible are either `final` or `sealed`, and other (legal) classes are freely extensible, unless I am missing something.

And unless I am mistaken, the `non-sealed` flag is not written to the classfiles, and is not reconstituted when reading from classfiles, so using `Symbol.isNonSealed()` is not easy, and we could only reconstitute the flag based on the fact that some of the supertypes are sealed, but the class itself is neither `final` nor `sealed.

It would be possible to implement the freely extensible check exactly as written, if desired, but it seems unnecessarily complex to me.

Reply:
@mcimadamore I can add a comment in this section of the code that refers to why we think that the current checks cover the reference in the spec to freely extensible class

Reply:
But, according to the spec, if I have:

```
sealed class Sup { }
non-sealed class Sub extends Sup { }
```

Then `Sub` is not freely extensible. Becaus, even though it is neither `sealed` nor `final`, its direct superclass is `sealed`. I agree that, in spirit, `Sub` can be extended, but that's not what the JLS seems to say?

Reply:
I believe the spec says it is freely extensible: a class which has a sealed direct super class or super interface is freely extensible iff the is non-sealed. Maybe I am missing something?
https://docs.oracle.com/javase/specs/jls/se22/html/jls-8.html#jls-8.1.1.2

Reply:
Ah, missed this:
```
 A class that has a sealed direct superclass or a sealed direct superinterface is freely extensible if and only if it is declared non-sealed. 
 ```

Reply:
Ok, this seems to be mostly a spec thing. We do have declarations (like records) that are not freely extensible, but do not need `final` or `sealed`. I suppose this is not a problem for javac because javac adds `final` to records and enums.